core/src/main/java/org/elasticsearch/cluster/metadata/IndexGraveyard.java

```diff
@@ -0,0 +1,436 @@
+            TOMBSTONE_PARSER.declareLong(Tombstone.Builder::deleteDateInMillis, new ParseField(DELETE_DATE_KEY));
+        }
+
+        static BiFunction<XContentParser, Void, Tombstone> getParser() {
+            return (p,c) -> TOMBSTONE_PARSER.apply(p, c).build();
+        }
+
+        private final Index index;
+        private final long deleteDateInMillis;
+
+        private Tombstone(final Index index, final long deleteDateInMillis) {
+            Objects.requireNonNull(index);
+            if (deleteDateInMillis < 0L) {
+                throw new IllegalArgumentException("invalid deleteDateInMillis [" + deleteDateInMillis + "]");
+            }
+            this.index = index;
+            this.deleteDateInMillis = deleteDateInMillis;
+        }
+
+        // create from stream
+        private Tombstone(StreamInput in) throws IOException {
+            index = new Index(in);
+            deleteDateInMillis = in.readLong();
+        }
+
+        /**
+         * The deleted index.
+         */
+        public Index getIndex() {
+            return index;
+        }
+
+        /**
+         * The date in milliseconds that the index deletion event occurred, used for logging/debugging.
+         */
+        public long getDeleteDateInMillis() {
+            return deleteDateInMillis;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            index.writeTo(out);
+            out.writeLong(deleteDateInMillis);
+        }
+
+        @Override
+        public Tombstone readFrom(final StreamInput in) throws IOException {
+            return new Tombstone(in);
+        }
+
+        @Override
+        public boolean equals(final Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            Tombstone that = (Tombstone) other;
+            if (index.equals(that.index) == false) {
+                return false;
+            }
+            if (deleteDateInMillis != that.deleteDateInMillis) {
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = index.hashCode();
+            result = 31 * result + Long.hashCode(deleteDateInMillis);
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "[index=" + index + ", deleteDateInMillis=" + deleteDateInMillis + "]";
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject();
+            builder.field(INDEX_KEY);
+            index.toXContent(builder, params);
+            builder.field(DELETE_DATE_KEY, deleteDateInMillis);
```

Comment:
It doesn't.  My point was only that, if we are writing a timeValueField to X-Content, then we need to also read the timeValueField from X-Content.  I'm using the `ObjectParser` to read from X-Content, which does not at the moment have anything for reading timeValueField.  I can add it though, if we deem this is the right way to go.

The point about the `Index.Builder` is another separate issue related to the use of `ObjectParser`.  Since we are using `ObjectParser` to read the Tombstone, it relies on the `Index`'s object parser, which writes to an `Index.Builder` in order to keep `Index` immutable.